SOSS/dms/soss_centroids.py

```diff
@@ -362,6 +435,860 @@ def get_uncontam_centroids_edgetrig(image, header=None, mask=None, poly_order=11
+    or below the trace.
+
+    :param ytrace: the trace y-position at each column, must have shape = (2048,).
+    :param subarray: the subarray for which to build a mask.
+    :param halfwidth: the size of the window to mask around the trace.
+    :param extend_below: if True mask all pixels above the trace.
+    :param extend_above: if True mask all pixels below the trace.
+
+    :type ytrace: array[float]
+    :type subarray: str
+    :type halfwidth: float
+    :type extend_below: bool
+    :type extend_above: bool
+
+    :returns: mask_trace - A mask that removes an area centered on the given trace positions.
+    :rtype: array[bool]
+    """
+
+    dimx = 2048
+
+    # Check the shape of the y-positions.
+    if np.shape(ytrace) != (dimx,):
+        msg = 'ytrace must have shape (2048,)'
+        raise ValueError(msg)
+
+    # Check the subarray value and set dimy accordingly.
+    if subarray == 'FULL':
+        dimy = 2048
+    elif subarray == 'SUBSTRIP96':
+        dimy = 96
+    elif subarray == 'SUBSTRIP256':
+        dimy = 256
+    else:
+        msg = 'Unknown subarray: {}'
+        raise ValueError(msg.format(subarray))
+
+    # Cannot both be True, that would mask everything.
+    if extend_below and extend_above:
+        msg = 'Only one of extend_below, extend_above should be used.'
+        raise ValueError(msg)
+
+    # Create a coordinate grid.
+    x = np.arange(dimx)
+    y = np.arange(dimy)
+    _, ygrid = np.meshgrid(x, y)
+
+    # Mask the pixels within a halfwidth of the trace center.
+    mask_trace = np.abs(ygrid - ytrace) < halfwidth
+
+    # If True mask all pixels below the trace center.
+    if extend_below:
+        mask_below = (ygrid - ytrace) < 0
+        mask_trace = mask_trace | mask_below
+
+    # If True mask all pixels above the trace center.
+    if extend_above:
+        mask_above = (ygrid - ytrace) >= 0
+        mask_trace = mask_trace | mask_above
+
+    return mask_trace
+
+
+def build_mask_order2_contaminated(ytrace_o1, ytrace_o3, subarray='SUBSTRIP256',
+                                   halfwidth_o1=25, halfwidth_o3=15, xlim=150):
+    """Build a mask that isolates the contaminated part of the order 2 trace.
+    This is done by masking the order 1 trace and averything below, the order
+    2 trace and everything above and all pixels blue-ward (to the right) of xlim.
+
+    :param ytrace_o1: y position of the order 1 trace at every column.
+    :param ytrace_o3: y position of the order 3 trace at every column.
+    :param subarray: the subarray for which to build a mask.
+    :param halfwidth_o1: the size of the window to mask around the order 1 trace.
+    :param halfwidth_o3: the size of the window to mask around the order 3 trace.
+    :param xlim: the boundary for masking pixels blue-ward (to the right).
+
+    :type ytrace_o1: array[float]
+    :type ytrace_o3: array[float]
+    :type subarray: str
+    :type halfwidth_o1: float
+    :type halfwidth_o3: float
+    :type xlim: float
+
```

Comment:
you can actually add a "raises" flag in the doc stringv

```python
:raises ValueError: if subarray not either 'FULL', 'SUBSTRIP96', 'SUBSTRIP256'
```

only really useful for others using this function - that they may need to catch this exception - in this case probably isn't that useful - but extra info is always good